File: juno/strategies/four_week_rule.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from decimal import Decimal

from juno import Advice, Candle, CandleMeta, indicators
from juno.constraints import Int
from juno.indicators import MA
from juno.inspect import Constructor, get_module_type
from juno.math import minmax

from .strategy import Signal, Strategy, ma_choices


@dataclass
class FourWeekRuleParams(Constructor):
    period: int = 28
    ma: str = "ema"
    ma_period: int = 14  # Normally half the period.

    def construct(self) -> FourWeekRule:
        return FourWeekRule(
            period=self.period,
            ma=self.ma,
            ma_period=self.ma_period,
        )


# Signals a long position when a candle close price goes above a highest four week close price.
# Signals a short position when a candle close price goes below a lowest four week close price.
# Signals liquidation when a candle close price crosses a two week moving average.
# Works with daily candles!
class FourWeekRule(Signal):
    @staticmethod
    def meta() -> Strategy.Meta:
        return Strategy.Meta(
            constraints={
                "period": Int(2, 100),
                "ma": ma_choices,
                "ma_period": Int(2, 100),
            }
        )

    _prices: deque[Decimal]
    _ma: MA
    _advice: Advice = Advice.NONE
    _t: int = 0
    _t1: int

    def __init__(
        self,
        period: int = 28,
        ma: str = "ema",
        ma_period: int = 14,  # Normally half the period.
    ) -> None:
        self._prices = deque(maxlen=period)
        self._ma = get_module_type(indicators, ma)(ma_period)
        self._t1 = period + 1

    @property
    def advice(self) -> Advice:
        return self._advice

    @property
    def maturity(self) -> int:
        return self._t1

    @property
    def mature(self) -> bool:
        return self._t >= self._t1

    def update(self, candle: Candle, _: CandleMeta) -> None:
        self._t = min(self._t + 1, self._t1)

        self._ma.update(candle.close)

        if self._t >= self._t1:
            lowest, highest = minmax(self._prices)
            if candle.close >= highest:
                self._advice = Advice.LONG
            elif candle.close <= lowest:
                self._advice = Advice.SHORT
            elif (self._advice is Advice.LONG and candle.close <= self._ma.value) or (
                self._advice is Advice.SHORT and candle.close >= self._ma.value
            ):
                self._advice = Advice.LIQUIDATE

        self._prices.append(candle.close)

File: juno_rs/strategies/four_week_rule.rs
use std::{cmp::min, collections::VecDeque};
use field_count::FieldCount;
use rand::{Rng, rngs::StdRng};
use crate::{
    indicators, math,
    strategies::{MidTrend, Strategy},
    Advice, Candle,
};

#[derive(Clone, FieldCount)]
#[repr(C)]
pub struct FourWeekRuleParams {
    pub period: u32,
    pub ma: u32,
    pub ma_period: u32,
    pub mid_trend_policy: u32,
}

pub struct FourWeekRule {
    mid_trend: MidTrend,
    prices: VecDeque<f64>,
    ma: Box<dyn indicators::MA>,
    advice: Advice,
    t: u32,
    t1: u32,
}

impl Strategy for FourWeekRule {
    type Params = FourWeekRuleParams;

    fn new(params: &Self::Params) -> Self {
        Self {
            mid_trend: MidTrend::new(params.mid_trend_policy),
            prices: VecDeque::with_capacity(params.period as usize),
            ma: indicators::ma_from_adler32(params.ma, params.ma_period),
            advice: Advice::None,
            t: 0,
            t1: params.period,
        }
    }

    fn update(&mut self, candle: &Candle) -> Advice {
        self.ma.update(candle.close);

        let mut advice = Advice::None;
        if self.t >= self.t1 {
            let (lowest, highest) = math::minmax(self.prices.iter());

            if candle.close >= highest {
                self.advice = Advice::Long;
            } else if candle.close <= lowest {
                self.advice = Advice::Short;
            } else if (self.advice == Advice::Long && candle.close <= self.ma.value())
                || (self.advice == Advice::Short && candle.close >= self.ma.value())
            {
                self.advice = Advice::Liquidate;
            }
            advice = self.mid_trend.update(self.advice);

            self.prices.pop_front();
        }

        self.prices.push_back(candle.close);
        self.t = min(self.t + 1, self.t1);
        advice
    }

    fn generate(rng: &mut StdRng) -> Self::Params {
        Self::Params {
            period: rng.gen_range(2, 100),
            ma: indicators::MA_CHOICES[rng.gen_range(0, indicators::MA_CHOICES.len())],
            ma_period: rng.gen_range(2, 100),
            mid_trend_policy: MidTrend::POLICY_IGNORE,
        }
    }
}
